File: lib/src/calib3d/stereo.dart
// Copyright (c) 2025, rainyl and all contributors. All rights reserved.
// Use of this source code is governed by a Apache-2.0 license
// that can be found in the LICENSE file.

// ignore_for_file: non_constant_identifier_names

library cv.calib3d;

import 'dart:ffi' as ffi;

import 'package:ffi/ffi.dart';

import '../core/base.dart';
import '../core/mat.dart';
import '../core/rect.dart';
import '../g/calib3d.g.dart' as cvg;

/// Class for computing stereo correspondence using the block matching algorithm,
/// introduced and contributed to OpenCV by K. Konolige.
///
/// This class implements a block matching algorithm for stereo correspondence,
/// which is used to compute disparity maps from stereo image pairs.
/// It provides methods to fine-tune parameters such as pre-filtering, texture thresholds,
/// uniqueness ratios, and regions of interest (ROIs) to optimize performance and accuracy.
///
/// https://docs.opencv.org/4.x/d9/dba/classcv_1_1StereoBM.html#details
class StereoBM extends CvStruct<cvg.StereoBM> {
  StereoBM._(cvg.StereoBMPtr ptr, [bool attach = true]) : super.fromPointer(ptr) {
    if (attach) {
      finalizer.attach(this, ptr.cast(), detach: this);
    }
  }

  factory StereoBM.fromPointer(cvg.StereoBMPtr ptr) => StereoBM._(ptr);

  /// Creates [StereoBM] object.
  ///
  /// https://docs.opencv.org/4.x/d9/dba/classcv_1_1StereoBM.html#a119436b6cb382e0895dd0fa58229ec17
  factory StereoBM.create({int numDisparities = 0, int blockSize = 21}) {
    final p = calloc<cvg.StereoBM>();
    cvRun(() => cvg.cv_StereoBM_create(numDisparities, blockSize, p, ffi.nullptr));
    return StereoBM._(p);
  }

  static final finalizer = OcvFinalizer<cvg.StereoBMPtr>(cvg.addresses.cv_StereoBM_close);

  int get preFilterCap => cvg.cv_StereoBM_getPreFilterCap(ref);
  set preFilterCap(int val) => cvg.cv_StereoBM_setPreFilterCap(ref, val);

  int get preFilterSize => cvg.cv_StereoBM_getPreFilterSize(ref);
  set preFilterSize(int val) => cvg.cv_StereoBM_setPreFilterSize(ref, val);

  int get preFilterType => cvg.cv_StereoBM_getPreFilterType(ref);
  set preFilterType(int val) => cvg.cv_StereoBM_setPreFilterType(ref, val);

  Rect get ROI1 => Rect.fromPointer(cvg.cv_StereoBM_getROI1(ref));
  set ROI1(Rect val) => cvg.cv_StereoBM_setROI1(ref, val.ref);

  Rect get ROI2 => Rect.fromPointer(cvg.cv_StereoBM_getROI2(ref));
  set ROI2(Rect val) => cvg.cv_StereoBM_setROI2(ref, val.ref);

  int get smallerBlockSize => cvg.cv_StereoBM_getSmallerBlockSize(ref);

  /// Warning: do nothing
  ///
  /// https://github.com/opencv/opencv/blob/75d9ac39643f8e08edf3299e47ffd1468e0c5196/modules/calib3d/src/stereobm.cpp#L1335
  set smallerBlockSize(int val) => cvg.cv_StereoBM_setSmallerBlockSize(ref, val);

  int get textureThreshold => cvg.cv_StereoBM_getTextureThreshold(ref);
  set textureThreshold(int val) => cvg.cv_StereoBM_setTextureThreshold(ref, val);

  int get uniquenessRatio => cvg.cv_StereoBM_getUniquenessRatio(ref);
  set uniquenessRatio(int val) => cvg.cv_StereoBM_setUniquenessRatio(ref, val);

  //  Public Member Functions inherited from cv::StereoMatcher

  Mat compute(Mat left, Mat right, {Mat? disparity}) {
    disparity ??= Mat.empty();
    cvRun(() => cvg.cv_StereoBM_compute(ref, left.ref, right.ref, disparity!.ref));
    return disparity;
  }

  int get blockSize => cvg.cv_StereoBM_getBlockSize(ref);
  set blockSize(int val) => cvg.cv_StereoBM_setBlockSize(ref, val);

  int get disp12MaxDiff => cvg.cv_StereoBM_getDisp12MaxDiff(ref);
  set disp12MaxDiff(int val) => cvg.cv_StereoBM_setDisp12MaxDiff(ref, val);

  int get minDisparity => cvg.cv_StereoBM_getMinDisparity(ref);
  set minDisparity(int val) => cvg.cv_StereoBM_setMinDisparity(ref, val);

  // int  getNumDisparities
  int get numDisparities => cvg.cv_StereoBM_getNumDisparities(ref);
  set numDisparities(int val) => cvg.cv_StereoBM_setNumDisparities(ref, val);

  // int  getSpeckleRange
  int get speckleRange => cvg.cv_StereoBM_getSpeckleRange(ref);
  set speckleRange(int val) => cvg.cv_StereoBM_setSpeckleRange(ref, val);

  // int  getSpeckleWindowSize
  int get speckleWindowSize => cvg.cv_StereoBM_getSpeckleWindowSize(ref);
  set speckleWindowSize(int val) => cvg.cv_StereoBM_setSpeckleWindowSize(ref, val);

  void dispose() {
    finalizer.detach(this);
    cvg.cv_StereoBM_close(ptr);
  }

  @override
  cvg.StereoBM get ref => ptr.ref;

  @override
  String toString() {
    return "StereoBM(address=0x${ptr.address.toRadixString(16)})";
  }

  /// Pre-filter types for the stereo matching algorithm.
  /// Uses normalized response for pre-filtering.
  static const int PREFILTER_NORMALIZED_RESPONSE = 0;

  /// Pre-filter types for the stereo matching algorithm.
  /// Uses the X-Sobel operator for pre-filtering.
  static const int PREFILTER_XSOBEL = 1;
}

/// The class implements the modified H. Hirschmuller algorithm [128] that differs from the original one as follows:
///
/// * By default, the algorithm is single-pass, which means that you consider only 5 directions instead of 8. Set mode=StereoSGBM::MODE_HH in createStereoSGBM to run the full variant of the algorithm but beware that it may consume a lot of memory.
/// * The algorithm matches blocks, not individual pixels. Though, setting blockSize=1 reduces the blocks to single pixels.
/// * Mutual information cost function is not implemented. Instead, a simpler Birchfield-Tomasi sub-pixel metric from [29] is used. Though, the color images are supported as well.
/// * Some pre- and post- processing steps from K. Konolige algorithm StereoBM are included, for example: pre-filtering (StereoBM::PREFILTER_XSOBEL type) and post-filtering (uniqueness check, quadratic interpolation and speckle filtering).
///
/// https://docs.opencv.org/4.x/d2/d85/classcv_1_1StereoSGBM.html#details
class StereoSGBM extends CvStruct<cvg.StereoSGBM> {
  StereoSGBM._(cvg.StereoSGBMPtr ptr, [bool attach = true]) : super.fromPointer(ptr) {
    if (attach) {
      finalizer.attach(this, ptr.cast(), detach: this);
    }
  }

  factory StereoSGBM.fromPointer(cvg.StereoSGBMPtr ptr) => StereoSGBM._(ptr);

  /// Creates StereoSGBM object.
  ///
  /// https://docs.opencv.org/4.x/d2/d85/classcv_1_1StereoSGBM.html#a177fd40dc0dae3a96059dbb3ad92a019
  factory StereoSGBM.create({
    int minDisparity = 0,
    int numDisparities = 16,
    int blockSize = 3,
    int P1 = 0,
    int P2 = 0,
    int disp12MaxDiff = 0,
    int preFilterCap = 0,
    int uniquenessRatio = 0,
    int speckleWindowSize = 0,
    int speckleRange = 0,
    int mode = MODE_SGBM,
  }) {
    final p = calloc<cvg.StereoSGBM>();
    cvRun(
      () => cvg.cv_StereoSGBM_create(
        minDisparity,
        numDisparities,
        blockSize,
        P1,
        P2,
        disp12MaxDiff,
        preFilterCap,
        uniquenessRatio,
        speckleWindowSize,
        speckleRange,
        mode,
        p,
        ffi.nullptr,
      ),
    );
    return StereoSGBM._(p);
  }

  static final finalizer = OcvFinalizer<cvg.StereoSGBMPtr>(cvg.addresses.cv_StereoSGBM_close);

  int get preFilterCap => cvg.cv_StereoSGBM_getPreFilterCap(ref);
  set preFilterCap(int val) => cvg.cv_StereoSGBM_setPreFilterCap(ref, val);

  int get mode => cvg.cv_StereoSGBM_getMode(ref);
  set mode(int val) => cvg.cv_StereoSGBM_setMode(ref, val);

  int get P1 => cvg.cv_StereoSGBM_getP1(ref);
  set P1(int val) => cvg.cv_StereoSGBM_setP1(ref, val);

  int get P2 => cvg.cv_StereoSGBM_getP2(ref);
  set P2(int val) => cvg.cv_StereoSGBM_setP2(ref, val);

  int get uniquenessRatio => cvg.cv_StereoSGBM_getUniquenessRatio(ref);
  set uniquenessRatio(int val) => cvg.cv_StereoSGBM_setUniquenessRatio(ref, val);

  //  Public Member Functions inherited from cv::StereoMatcher

  Mat compute(Mat left, Mat right, {Mat? disparity}) {
    disparity ??= Mat.empty();
    cvRun(() => cvg.cv_StereoSGBM_compute(ref, left.ref, right.ref, disparity!.ref));
    return disparity;
  }

  int get blockSize => cvg.cv_StereoSGBM_getBlockSize(ref);
  set blockSize(int val) => cvg.cv_StereoSGBM_setBlockSize(ref, val);

  int get disp12MaxDiff => cvg.cv_StereoSGBM_getDisp12MaxDiff(ref);
  set disp12MaxDiff(int val) => cvg.cv_StereoSGBM_setDisp12MaxDiff(ref, val);

  int get minDisparity => cvg.cv_StereoSGBM_getMinDisparity(ref);
  set minDisparity(int val) => cvg.cv_StereoSGBM_setMinDisparity(ref, val);

  // int  getNumDisparities
  int get numDisparities => cvg.cv_StereoSGBM_getNumDisparities(ref);
  set numDisparities(int val) => cvg.cv_StereoSGBM_setNumDisparities(ref, val);

  // int  getSpeckleRange
  int get speckleRange => cvg.cv_StereoSGBM_getSpeckleRange(ref);
  set speckleRange(int val) => cvg.cv_StereoSGBM_setSpeckleRange(ref, val);

  // int  getSpeckleWindowSize
  int get speckleWindowSize => cvg.cv_StereoSGBM_getSpeckleWindowSize(ref);
  set speckleWindowSize(int val) => cvg.cv_StereoSGBM_setSpeckleWindowSize(ref, val);

  void dispose() {
    finalizer.detach(this);
    cvg.cv_StereoSGBM_close(ptr);
  }

  @override
  cvg.StereoSGBM get ref => ptr.ref;

  @override
  String toString() {
    return "StereoSGBM(address=0x${ptr.address.toRadixString(16)})";
  }

  static const int MODE_SGBM = 0;
  static const int MODE_HH = 1;
  static const int MODE_SGBM_3WAY = 2;
  static const int MODE_HH4 = 3;
}

File: packages/dartcv/src/dartcv/calib3d/stereo.cpp
//
// Created by rainy on 2025/8/12.
//

#include "stereo.h"

// StereoBM
CvStatus* cv_StereoBM_create(
    int numDisparities, int blockSize, StereoBM* rval, CvCallback_0 callback
) {
    BEGIN_WRAP
    *rval = {new cv::Ptr<cv::StereoBM>(cv::StereoBM::create(numDisparities, blockSize))};
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

void cv_StereoBM_close(StereoBMPtr self) {
    self->ptr->reset();
    CVD_FREE(self);
}

// Gets the current truncation value for prefiltered pixels.
// virtual int  getPreFilterCap () const =0
int cv_StereoBM_getPreFilterCap(StereoBM self) {
    return (CVDEREF(self))->getPreFilterCap();
}

// Gets the current size of the pre-filter kernel.
// virtual int  getPreFilterSize () const =0
int cv_StereoBM_getPreFilterSize(StereoBM self) {
    return (CVDEREF(self))->getPreFilterSize();
}

// Gets the type of pre-filtering currently used in the algorithm.
// virtual int  getPreFilterType () const =0
int cv_StereoBM_getPreFilterType(StereoBM self) {
    return (CVDEREF(self))->getPreFilterType();
}

// Gets the current Region of Interest (ROI) for the left image.
// virtual Rect  getROI1 () const =0
CvRect* cv_StereoBM_getROI1(StereoBM self) {
    auto rect = (CVDEREF(self))->getROI1();
    return new CvRect{rect.x, rect.y, rect.width, rect.height};
}

// Gets the current Region of Interest (ROI) for the right image.
// virtual Rect  getROI2 () const =0
CvRect* cv_StereoBM_getROI2(StereoBM self) {
    auto rect = (CVDEREF(self))->getROI2();
    return new CvRect{rect.x, rect.y, rect.width, rect.height};
}

// Gets the current size of the smaller block used for texture check.
// virtual int  getSmallerBlockSize () const =0
int cv_StereoBM_getSmallerBlockSize(StereoBM self) {
    return (CVDEREF(self))->getSmallerBlockSize();
}

// Gets the current texture threshold value.
// virtual int  getTextureThreshold () const =0
int cv_StereoBM_getTextureThreshold(StereoBM self) {
    return (CVDEREF(self))->getTextureThreshold();
}

// Gets the current uniqueness ratio value.
// virtual int  getUniquenessRatio () const =0
int cv_StereoBM_getUniquenessRatio(StereoBM self) {
    return (CVDEREF(self))->getUniquenessRatio();
}

// Sets the truncation value for prefiltered pixels.
// virtual void  setPreFilterCap (int preFilterCap)=0
void cv_StereoBM_setPreFilterCap(StereoBM self, int preFilterCap) {
    (CVDEREF(self))->setPreFilterCap(preFilterCap);
}

// Sets the size of the pre-filter kernel.
// virtual void  setPreFilterSize (int preFilterSize)=0
void cv_StereoBM_setPreFilterSize(StereoBM self, int preFilterSize) {
    (CVDEREF(self))->setPreFilterSize(preFilterSize);
}

// Sets the type of pre-filtering used in the algorithm.
// virtual void  setPreFilterType (int preFilterType)=0
void cv_StereoBM_setPreFilterType(StereoBM self, int preFilterType) {
    (CVDEREF(self))->setPreFilterType(preFilterType);
}

// Sets the Region of Interest (ROI) for the left image.
// virtual void  setROI1 (Rect roi1)=0
void cv_StereoBM_setROI1(StereoBM self, CvRect roi1) {
    auto rect = cv::Rect(roi1.x, roi1.y, roi1.width, roi1.height);
    (CVDEREF(self))->setROI1(rect);
}

// Sets the Region of Interest (ROI) for the right image.
// virtual void  setROI2 (Rect roi2)=0
void cv_StereoBM_setROI2(StereoBM self, CvRect roi2) {
    auto rect = cv::Rect(roi2.x, roi2.y, roi2.width, roi2.height);
    (CVDEREF(self))->setROI2(rect);
}

// Sets the size of the smaller block used for texture check.
// virtual void  setSmallerBlockSize (int blockSize)=0
void cv_StereoBM_setSmallerBlockSize(StereoBM self, int blockSize) {
    (CVDEREF(self))->setSmallerBlockSize(blockSize);
}

// Sets the threshold for filtering low-texture regions.
// virtual void  setTextureThreshold (int textureThreshold)=0
void cv_StereoBM_setTextureThreshold(StereoBM self, int textureThreshold) {
    (CVDEREF(self))->setTextureThreshold(textureThreshold);
}

// Sets the uniqueness ratio for filtering ambiguous matches.
// virtual void  setUniquenessRatio (int uniquenessRatio)=0
void cv_StereoBM_setUniquenessRatio(StereoBM self, int uniquenessRatio) {
    (CVDEREF(self))->setUniquenessRatio(uniquenessRatio);
}

// Computes disparity map for the specified stereo pair.
// virtual void  compute (InputArray left, InputArray right, OutputArray disparity)=0
CvStatus* cv_StereoBM_compute(StereoBM self, MatIn left, MatIn right, MatOut disparity) {
    BEGIN_WRAP(CVDEREF(self))->compute(CVDEREF(left), CVDEREF(right), CVDEREF(disparity));
    END_WRAP
}

// virtual int  getBlockSize () const =0
int cv_StereoBM_getBlockSize(StereoBM self) {
    return (CVDEREF(self))->getBlockSize();
}

// virtual int  getDisp12MaxDiff () const =0
int cv_StereoBM_getDisp12MaxDiff(StereoBM self) {
    return (CVDEREF(self))->getDisp12MaxDiff();
}

// virtual int  getMinDisparity () const =0
int cv_StereoBM_getMinDisparity(StereoBM self) {
    return (CVDEREF(self))->getMinDisparity();
}

// virtual int  getNumDisparities () const =0
int cv_StereoBM_getNumDisparities(StereoBM self) {
    return (CVDEREF(self))->getNumDisparities();
}

// virtual int  getSpeckleRange () const =0
int cv_StereoBM_getSpeckleRange(StereoBM self) {
    return (CVDEREF(self))->getSpeckleRange();
}

// virtual int  getSpeckleWindowSize () const =0
int cv_StereoBM_getSpeckleWindowSize(StereoBM self) {
    return (CVDEREF(self))->getSpeckleWindowSize();
}

// virtual void  setBlockSize (int blockSize)=0
void cv_StereoBM_setBlockSize(StereoBM self, int blockSize) {
    (CVDEREF(self))->setBlockSize(blockSize);
}

// virtual void  setDisp12MaxDiff (int disp12MaxDiff)=0
void cv_StereoBM_setDisp12MaxDiff(StereoBM self, int disp12MaxDiff) {
    (CVDEREF(self))->setDisp12MaxDiff(disp12MaxDiff);
}

// virtual void  setMinDisparity (int minDisparity)=0
void cv_StereoBM_setMinDisparity(StereoBM self, int minDisparity) {
    (CVDEREF(self))->setMinDisparity(minDisparity);
}

// virtual void  setNumDisparities (int numDisparities)=0
void cv_StereoBM_setNumDisparities(StereoBM self, int numDisparities) {
    (CVDEREF(self))->setNumDisparities(numDisparities);
}

// virtual void  setSpeckleRange (int speckleRange)=0
void cv_StereoBM_setSpeckleRange(StereoBM self, int speckleRange) {
    (CVDEREF(self))->setSpeckleRange(speckleRange);
}

// virtual void  setSpeckleWindowSize (int speckleWindowSize)=0
void cv_StereoBM_setSpeckleWindowSize(StereoBM self, int speckleWindowSize) {
    (CVDEREF(self))->setSpeckleWindowSize(speckleWindowSize);
}

// StereoSGBM
CvStatus* cv_StereoSGBM_create(
    int minDisparity,
    int numDisparities,
    int blockSize,
    int P1,
    int P2,
    int disp12MaxDiff,
    int preFilterCap,
    int uniquenessRatio,
    int speckleWindowSize,
    int speckleRange,
    int mode,
    StereoSGBM* rval,
    CvCallback_0 callback
) {
    BEGIN_WRAP
    *rval = {new cv::Ptr<cv::StereoSGBM>(cv::StereoSGBM::create(
        minDisparity,
        numDisparities,
        blockSize,
        P1,
        P2,
        disp12MaxDiff,
        preFilterCap,
        uniquenessRatio,
        speckleWindowSize,
        speckleRange,
        mode
    ))};
    if (callback != nullptr) {
        callback();
    }
    END_WRAP
}

void cv_StereoSGBM_close(StereoSGBMPtr self) {
    self->ptr->reset();
    CVD_FREE(self);
}

// virtual int getMode () const =0
int cv_StereoSGBM_getMode(StereoSGBM self) {
    return (CVDEREF(self))->getMode();
}

// virtual int getP1 () const =0
int cv_StereoSGBM_getP1(StereoSGBM self) {
    return (CVDEREF(self))->getP1();
}

// virtual int getP2 () const =0
int cv_StereoSGBM_getP2(StereoSGBM self) {
    return (CVDEREF(self))->getP2();
}

// virtual int getPreFilterCap () const =0
int cv_StereoSGBM_getPreFilterCap(StereoSGBM self) {
    return (CVDEREF(self))->getPreFilterCap();
}

// virtual int getUniquenessRatio () const =0
int cv_StereoSGBM_getUniquenessRatio(StereoSGBM self) {
    return (CVDEREF(self))->getUniquenessRatio();
}

// virtual void setMode (int mode)=0
void cv_StereoSGBM_setMode(StereoSGBM self, int mode) {
    (CVDEREF(self))->setMode(mode);
}

// virtual void setP1 (int P1)=0
void cv_StereoSGBM_setP1(StereoSGBM self, int P1) {
    (CVDEREF(self))->setP1(P1);
}

// virtual void setP2 (int P2)=0
void cv_StereoSGBM_setP2(StereoSGBM self, int P2) {
    (CVDEREF(self))->setP2(P2);
}

// virtual void setPreFilterCap (int preFilterCap)=0
void cv_StereoSGBM_setPreFilterCap(StereoSGBM self, int preFilterCap) {
    (CVDEREF(self))->setPreFilterCap(preFilterCap);
}

// virtual void setUniquenessRatio (int uniquenessRatio)=0
void cv_StereoSGBM_setUniquenessRatio(StereoSGBM self, int uniquenessRatio) {
    (CVDEREF(self))->setUniquenessRatio(uniquenessRatio);
}

// Computes disparity map for the specified stereo pair.
// virtual void compute (InputArray left, InputArray right, OutputArray disparity)=0
CvStatus* cv_StereoSGBM_compute(StereoSGBM self, MatIn left, MatIn right, MatOut disparity) {
    BEGIN_WRAP(CVDEREF(self))->compute(CVDEREF(left), CVDEREF(right), CVDEREF(disparity));
    END_WRAP
}

// virtual int getBlockSize () const =0
int cv_StereoSGBM_getBlockSize(StereoSGBM self) {
    return (CVDEREF(self))->getBlockSize();
}

// virtual int getDisp12MaxDiff () const =0
int cv_StereoSGBM_getDisp12MaxDiff(StereoSGBM self) {
    return (CVDEREF(self))->getDisp12MaxDiff();
}

// virtual int getMinDisparity () const =0
int cv_StereoSGBM_getMinDisparity(StereoSGBM self) {
    return (CVDEREF(self))->getMinDisparity();
}

// virtual int getNumDisparities () const =0
int cv_StereoSGBM_getNumDisparities(StereoSGBM self) {
    return (CVDEREF(self))->getNumDisparities();
}

// virtual int getSpeckleRange () const =0
int cv_StereoSGBM_getSpeckleRange(StereoSGBM self) {
    return (CVDEREF(self))->getSpeckleRange();
}

// virtual int getSpeckleWindowSize () const =0
int cv_StereoSGBM_getSpeckleWindowSize(StereoSGBM self) {
    return (CVDEREF(self))->getSpeckleWindowSize();
}

// virtual void setBlockSize (int blockSize)=0
void cv_StereoSGBM_setBlockSize(StereoSGBM self, int blockSize) {
    (CVDEREF(self))->setBlockSize(blockSize);
}

// virtual void setDisp12MaxDiff (int disp12MaxDiff)=0
void cv_StereoSGBM_setDisp12MaxDiff(StereoSGBM self, int disp12MaxDiff) {
    (CVDEREF(self))->setDisp12MaxDiff(disp12MaxDiff);
}

// virtual void setMinDisparity (int minDisparity)=0
void cv_StereoSGBM_setMinDisparity(StereoSGBM self, int minDisparity) {
    (CVDEREF(self))->setMinDisparity(minDisparity);
}

// virtual void setNumDisparities (int numDisparities)=0
void cv_StereoSGBM_setNumDisparities(StereoSGBM self, int numDisparities) {
    (CVDEREF(self))->setNumDisparities(numDisparities);
}

// virtual void setSpeckleRange (int speckleRange)=0
void cv_StereoSGBM_setSpeckleRange(StereoSGBM self, int speckleRange) {
    (CVDEREF(self))->setSpeckleRange(speckleRange);
}

// virtual void setSpeckleWindowSize (int speckleWindowSize)=0
void cv_StereoSGBM_setSpeckleWindowSize(StereoSGBM self, int speckleWindowSize) {
    (CVDEREF(self))->setSpeckleWindowSize(speckleWindowSize);
}
